From b33f518c5e3cd4a4514df71920a665f69c21722a Mon Sep 17 00:00:00 2001
From: Aishwarya R <aisr@qti.qualcomm.com>
Date: Wed, 4 Jun 2025 12:10:08 +0530
Subject: [PATCH 3/4] UPSTREAM:DP:wifi: mac80211: Enhance TX stats reporting
 per  TID specific with drop reasons

Improve the TX statistics reporting in the mac80211 subsystem by adding detailed
counters for various TX packet reasons and drop statistics for specific TIDs.

Below is the debugfs file format will be available for TID 0 - 15 with
all packet reasons and drop reasons

$cat sys/kernel/debug/ieee80211/phy00/netdev:wlan1/txrx_stats_reason
		MAC80211 TX STATS
 TID	 packets  bytes  eth_pkts  eth_bytes  fast_tx_pkts  fast_tx_bytes  nwifi_pkts  nwifi_bytes mon_pkts  mon_bytes  mc_pkts  mc_bytes
 0        616      57652    616        57652      0            0            0          0          0          0          149        21497
 1        0        0        0          0          0            0            0          0          0          0          0          0
 2        0        0        0          0          0            0            0          0          0          0          0          0
 3        0        0        0          0          0            0            0          0          0          0          0          0
 4        0        0        0          0          0            0            0          0          0          0          0          0
 5        0        0        0          0          0            0            0          0          0          0          0          0
 6        5        294      5          294        0            0            0          0          0          0          4          184
....
15       0        0        0            0              0              0              0                0          0

                MAC80211 TX DROP STATS
TID 0   5: 3,
TID 1
TID 2
TID 3

....
TID15

Patch-dependency:none
Signed-off-by: Girish Kumar M P <gmp@qti.qualcomm.com>
Signed-off-by: Aishwarya R <aisr@qti.qualcomm.com>
---
 include/net/mac80211.h |   1 +
 net/mac80211/drop.h    |  31 +++++++
 net/mac80211/iface.c   |   8 ++
 net/mac80211/tx.c      | 200 ++++++++++++++++++++++++++++++++++-------
 net/mac80211/wep.c     |   1 +
 net/mac80211/wpa.c     |   8 +-
 6 files changed, 216 insertions(+), 33 deletions(-)

diff --git a/include/net/mac80211.h b/include/net/mac80211.h
index f0c6aa1..f688a89 100644
--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -3094,6 +3094,7 @@ struct ieee80211_hw {
 	u32 dbg_mask;
 	u8 medium_busy;
 	bool perf_mode;
+	bool tid_stats_disable;
 };
 
 static inline bool _ieee80211_hw_check(struct ieee80211_hw *hw,
diff --git a/net/mac80211/drop.h b/net/mac80211/drop.h
index 08d8c57..8fe0aff 100644
--- a/net/mac80211/drop.h
+++ b/net/mac80211/drop.h
@@ -144,6 +144,37 @@ enum rx_drop_reasons {
 	RX_DROP_REASON_MAX
 };
 
+enum _tx_pkt_type {
+	TX_ETH_PKT,
+	TX_NWIFI_PKT,
+	TX_MONITOR_PKT,
+};
+
+enum tx_drop_reasons {
+	TX_DROP_SKB_ALLOC_FAIL,
+	TX_DROP_SKB_QUEUE_FULL,
+	TX_DROP_SKB_SANITY_CHECK_FAIL,
+	TX_DROP_SKB_RESIZE_FAIL,
+	TX_DROP_TX_SKB_FIXUP_FAIL,
+	TX_DROP_NOT_ASSOC_STA,
+	TX_DROP_KEY_FAIL,
+	TX_DROP_STA_MAX_BUFFER,
+	TX_DROP_SDATA_STATE,
+	TX_DROP_UNAUTHORIZED_STA,
+	TX_DROP_QUEUE_PURGE,
+	TX_DROP_MISC,
+	TX_DROP_BAD_CHANCTX_CONF,
+	TX_DROP_BAD_LINK_CONF,
+	TX_DROP_UNKNOWN_VIF_TYPE,
+	TX_DROP_UNAUTH_PORT,
+	TX_DROP_BAD_LINK_ID,
+	TX_DROP_HW_CHECK_FAIL,
+	TX_DROP_RA_STA_FAIL,
+	TX_DROP_WEP,
+	TX_DROP_REASON_MAX
+};
+
+static inline void ieee80211_tx_drop_stats(struct ieee80211_sub_if_data *sdata, u8 tid, u32 reason);
 #define RX_RES_IS_UNUSABLE(result)	\
 	(((__force u32)(result) & SKB_DROP_REASON_SUBSYS_MASK) == ___RX_DROP_UNUSABLE)
 
diff --git a/net/mac80211/iface.c b/net/mac80211/iface.c
index 9d75f99..5662e74 100644
--- a/net/mac80211/iface.c
+++ b/net/mac80211/iface.c
@@ -2352,6 +2352,7 @@ int ieee80211_if_add(struct ieee80211_local *local, const char *name,
 				GFP_KERNEL);
 		if (!sdata)
 			return -ENOMEM;
+
 		wdev = &sdata->wdev;
 
 		sdata->dev = NULL;
@@ -2438,6 +2439,12 @@ int ieee80211_if_add(struct ieee80211_local *local, const char *name,
 
 	ieee80211_sdata_init(local, sdata);
 
+	local->hw.tid_stats_disable = true;
+
+	sdata->txrx_stats = alloc_percpu_gfp(struct pcpu_txrx_stats, GFP_KERNEL);
+	if (!sdata->txrx_stats)
+		return -ENOMEM;
+
 	ieee80211_init_frag_cache(&sdata->frags);
 
 	INIT_LIST_HEAD(&sdata->key_list);
@@ -2537,6 +2544,7 @@ void ieee80211_if_remove(struct ieee80211_sub_if_data *sdata)
 
 	cfg80211_unregister_wdev(&sdata->wdev);
 
+	free_percpu(sdata->txrx_stats);
 	if (!sdata->dev) {
 		ieee80211_teardown_sdata(sdata);
 		kfree(sdata);
diff --git a/net/mac80211/tx.c b/net/mac80211/tx.c
index 9851790..ff02943 100644
--- a/net/mac80211/tx.c
+++ b/net/mac80211/tx.c
@@ -60,6 +60,55 @@ static inline void ieee80211_tx_stats(struct net_device *dev, u32 len)
 }
 /* misc utils */
 
+static inline void ieee80211_tid_classifier(struct sk_buff *skb,
+					    struct ieee80211_sub_if_data *sdata,
+					    bool fast_tx,
+					    u8 packet_type)
+{
+	struct ethhdr *ehdr = (struct ethhdr *)skb->data;
+	unsigned char *ra = ehdr->h_dest;
+	bool multicast = is_multicast_ether_addr(ra);
+	struct pcpu_txrx_stats *txrx_stats = this_cpu_ptr(sdata->txrx_stats);
+	struct ieee80211_tx_info *info = IEEE80211_SKB_CB(skb);
+	u8 tid;
+
+	info->tid = skb->priority & IEEE80211_QOS_CTL_TID_MASK;
+	tid = info->tid;
+
+	u64_stats_update_begin(&txrx_stats->syncp);
+
+	if (fast_tx) {
+		txrx_stats->tid_stats[tid].fast_tx_pkts++;
+		txrx_stats->tid_stats[tid].fast_tx_pkts_bytes += skb->len;
+		u64_stats_update_end(&txrx_stats->syncp);
+		return;
+	}
+
+	switch (packet_type) {
+	case TX_ETH_PKT:
+		txrx_stats->tid_stats[tid].tx_eth_pkts++;
+		txrx_stats->tid_stats[tid].tx_eth_pkts_bytes += skb->len;
+		break;
+	case TX_NWIFI_PKT:
+		txrx_stats->tid_stats[tid].tx_nwifi_pkts++;
+		txrx_stats->tid_stats[tid].tx_nwifi_pkts_bytes += skb->len;
+		break;
+	case TX_MONITOR_PKT:
+		txrx_stats->tid_stats[tid].tx_monitor_pkts++;
+		txrx_stats->tid_stats[tid].tx_monitor_pkts_bytes += skb->len;
+		break;
+	default:
+		break;
+	}
+	
+	if (multicast) {
+		txrx_stats->tid_stats[tid].tx_multicast_pkts++;
+		txrx_stats->tid_stats[tid].tx_multicast_pkts_bytes += skb->len;
+	}
+
+	u64_stats_update_end(&txrx_stats->syncp);
+}
+
 static __le16 ieee80211_duration(struct ieee80211_tx_data *tx,
 				 struct sk_buff *skb, int group_addr,
 				 int next_frag_len)
@@ -315,7 +364,7 @@ ieee80211_tx_h_check_assoc(struct ieee80211_tx_data *tx)
 	if (unlikely(test_bit(SCAN_SW_SCANNING, &tx->local->scanning)) &&
 	    test_bit(SDATA_STATE_OFFCHANNEL, &tx->sdata->state) &&
 	    !ieee80211_is_probe_req(hdr->frame_control) &&
-	    !ieee80211_is_any_nullfunc(hdr->frame_control))
+	    !ieee80211_is_any_nullfunc(hdr->frame_control)) {
 		/*
 		 * When software scanning only nullfunc frames (to notify
 		 * the sleep state to the AP) and probe requests (for the
@@ -327,7 +376,9 @@ ieee80211_tx_h_check_assoc(struct ieee80211_tx_data *tx)
 		 *
 		 * http://article.gmane.org/gmane.linux.kernel.wireless.general/30089
 		 */
+		ieee80211_tx_drop_stats(tx->sdata, info->tid, TX_DROP_SDATA_STATE);
 		return TX_DROP;
+	}
 
 	if (tx->sdata->vif.type == NL80211_IFTYPE_OCB)
 		return TX_CONTINUE;
@@ -347,6 +398,7 @@ ieee80211_tx_h_check_assoc(struct ieee80211_tx_data *tx)
 				   hdr->addr1);
 #endif
 			I802_DEBUG_INC(tx->local->tx_handlers_drop_not_assoc);
+			ieee80211_tx_drop_stats(tx->sdata, info->tid, TX_DROP_NOT_ASSOC_STA);
 			return TX_DROP;
 		}
 	} else if (unlikely(ieee80211_is_data(hdr->frame_control) &&
@@ -355,6 +407,7 @@ ieee80211_tx_h_check_assoc(struct ieee80211_tx_data *tx)
 		 * No associated STAs - no need to send multicast
 		 * frames.
 		 */
+		ieee80211_tx_drop_stats(tx->sdata, info->tid, TX_DROP_UNAUTHORIZED_STA);
 		return TX_DROP;
 	}
 
@@ -385,6 +438,7 @@ static void purge_old_ps_buffers(struct ieee80211_local *local)
 		skb = skb_dequeue(&ps->bc_buf);
 		if (skb) {
 			purged++;
+			ieee80211_tx_drop_stats(sdata, 0, TX_DROP_QUEUE_PURGE);
 			ieee80211_free_txskb(&local->hw, skb);
 		}
 		total += skb_queue_len(&ps->bc_buf);
@@ -402,6 +456,7 @@ static void purge_old_ps_buffers(struct ieee80211_local *local)
 			total += skb_queue_len(&sta->ps_tx_buf[ac]);
 			if (skb) {
 				purged++;
+				ieee80211_tx_drop_stats(sdata, 0, TX_DROP_QUEUE_PURGE);
 				ieee80211_free_txskb(&local->hw, skb);
 				break;
 			}
@@ -468,6 +523,7 @@ ieee80211_tx_h_multicast_ps_buf(struct ieee80211_tx_data *tx)
 	if (skb_queue_len(&ps->bc_buf) >= AP_MAX_BC_BUFFER) {
 		ps_dbg(tx->sdata,
 		       "BC TX buffer full - dropping the oldest frame\n");
+		ieee80211_tx_drop_stats(tx->sdata, info->tid, TX_DROP_STA_MAX_BUFFER);
 		ieee80211_free_txskb(&tx->local->hw, skb_dequeue(&ps->bc_buf));
 	} else
 		tx->local->total_ps_buffered++;
@@ -539,6 +595,7 @@ ieee80211_tx_h_unicast_ps_buf(struct ieee80211_tx_data *tx)
 			ps_dbg(tx->sdata,
 			       "STA %pM TX buffer for AC %d full - dropping oldest frame\n",
 			       sta->sta.addr, ac);
+			ieee80211_tx_drop_stats(tx->sdata, info->tid, TX_DROP_STA_MAX_BUFFER);
 			ieee80211_free_txskb(&local->hw, old);
 		} else
 			tx->local->total_ps_buffered++;
@@ -687,14 +744,17 @@ ieee80211_tx_h_select_key(struct ieee80211_tx_data *tx)
 
 		if (unlikely(tx->key && tx->key->flags & KEY_FLAG_TAINTED &&
 			     !ieee80211_is_deauth(hdr->frame_control)) &&
-			     tx->skb->protocol != tx->sdata->control_port_protocol)
+			     tx->skb->protocol != tx->sdata->control_port_protocol) {
+			ieee80211_tx_drop_stats(tx->sdata, info->tid, TX_DROP_KEY_FAIL);
 			return TX_DROP;
+		}
 
 		if (!skip_hw && tx->key &&
 		    tx->key->flags & KEY_FLAG_UPLOADED_TO_HARDWARE)
 			info->control.hw_key = &tx->key->conf;
 	} else if (ieee80211_is_data_present(hdr->frame_control) && tx->sta &&
 		   test_sta_flag(tx->sta, WLAN_STA_USES_ENCRYPTION)) {
+		ieee80211_tx_drop_stats(tx->sdata, info->tid, TX_DROP_MISC);
 		return TX_DROP;
 	}
 
@@ -936,8 +996,10 @@ static int ieee80211_fragment(struct ieee80211_tx_data *tx,
 				    frag_threshold +
 				    IEEE80211_ENCRYPT_HEADROOM +
 				    IEEE80211_ENCRYPT_TAILROOM);
-		if (!tmp)
+		if (!tmp) {
+			ieee80211_tx_drop_stats(tx->sdata, info->tid, TX_DROP_SKB_ALLOC_FAIL);
 			return -ENOMEM;
+		}
 
 		__skb_queue_tail(&tx->skbs, tmp);
 
@@ -1267,8 +1329,10 @@ ieee80211_tx_prepare(struct ieee80211_sub_if_data *sdata,
 	} else {
 		if (sdata->vif.type == NL80211_IFTYPE_AP_VLAN) {
 			tx->sta = rcu_dereference(sdata->u.vlan.sta);
-			if (!tx->sta && sdata->wdev.use_4addr)
+			if (!tx->sta && sdata->wdev.use_4addr) {
+				ieee80211_tx_drop_stats(sdata, info->tid, TX_DROP_MISC);
 				return TX_DROP;
+			}
 		} else if (tx->sdata->control_port_protocol == tx->skb->protocol) {
 			tx->sta = sta_info_get_bss(sdata, hdr->addr1);
 		}
@@ -1749,6 +1813,7 @@ static bool ieee80211_tx_frags(struct ieee80211_local *local,
 #ifdef CPTCFG_MAC80211_VERBOSE_DEBUG
 		if (WARN_ON_ONCE(q >= local->hw.queues)) {
 			__skb_unlink(skb, skbs);
+			ieee80211_tx_drop_stats(sta->sdata, info->tid,  TX_DROP_SKB_QUEUE_FULL);
 			ieee80211_free_txskb(&local->hw, skb);
 			continue;
 		}
@@ -2038,6 +2103,7 @@ static bool ieee80211_tx(struct ieee80211_sub_if_data *sdata,
 	bool result = true;
 
 	if (unlikely(skb->len < 10)) {
+		ieee80211_tx_drop_stats(sdata, info->tid, TX_DROP_SKB_SANITY_CHECK_FAIL);
 		dev_kfree_skb(skb);
 		return true;
 	}
@@ -2138,6 +2204,7 @@ void ieee80211_xmit(struct ieee80211_sub_if_data *sdata,
 	headroom = max_t(int, 0, headroom);
 
 	if (ieee80211_skb_resize(sdata, skb, headroom, encrypt)) {
+		ieee80211_tx_drop_stats(sdata, info->tid, TX_DROP_SKB_RESIZE_FAIL);
 		ieee80211_free_txskb(&local->hw, skb);
 		return;
 	}
@@ -2393,22 +2460,32 @@ netdev_tx_t ieee80211_monitor_start_xmit(struct sk_buff *skb,
 	struct cfg80211_chan_def *chandef;
 	u16 len_rthdr;
 	int hdrlen;
+	bool tid_stats_disable = local->hw.tid_stats_disable;
 
 #ifdef CPTCFG_MAC80211_ATHMEMDEBUG
 	ath_update_alloc(skb, skb->truesize, __LINE__, __func__, 1);
 #endif
-
 	sdata = IEEE80211_DEV_TO_SUB_IF(dev);
-	if (unlikely(!ieee80211_sdata_running(sdata)))
-		goto fail;
 
 	memset(info, 0, sizeof(*info));
 	info->flags = IEEE80211_TX_CTL_REQ_TX_STATUS |
 		      IEEE80211_TX_CTL_INJECTED;
 
+	if (!tid_stats_disable) {
+		skb->priority = cfg80211_classify8021d(skb, NULL);
+		ieee80211_tid_classifier(skb, sdata, false, TX_MONITOR_PKT);
+	}
+
+	if (unlikely(!ieee80211_sdata_running(sdata))) {
+		ieee80211_tx_drop_stats(sdata, info->tid, TX_DROP_SDATA_STATE);
+		goto fail;
+	}
+
 	/* Sanity-check the length of the radiotap header */
-	if (!ieee80211_validate_radiotap_len(skb))
+	if (!ieee80211_validate_radiotap_len(skb)) {
+		ieee80211_tx_drop_stats(sdata, info->tid, TX_DROP_SKB_SANITY_CHECK_FAIL);
 		goto fail;
+	}
 
 	/* we now know there is a radiotap header with a length we can use */
 	len_rthdr = ieee80211_get_radiotap_len(skb->data);
@@ -2427,14 +2504,18 @@ netdev_tx_t ieee80211_monitor_start_xmit(struct sk_buff *skb,
 	skb_set_network_header(skb, len_rthdr);
 	skb_set_transport_header(skb, len_rthdr);
 
-	if (skb->len < len_rthdr + 2)
+	if (skb->len < len_rthdr + 2) {
+		ieee80211_tx_drop_stats(sdata, info->tid, TX_DROP_SKB_SANITY_CHECK_FAIL);
 		goto fail;
+	}
 
 	hdr = (struct ieee80211_hdr *)(skb->data + len_rthdr);
 	hdrlen = ieee80211_hdrlen(hdr->frame_control);
 
-	if (skb->len < len_rthdr + hdrlen)
+	if (skb->len < len_rthdr + hdrlen) {
+		ieee80211_tx_drop_stats(sdata, info->tid, TX_DROP_SKB_SANITY_CHECK_FAIL);
 		goto fail;
+	}
 
 	/*
 	 * Initialize skb->protocol if the injected frame is a data frame
@@ -2485,8 +2566,10 @@ netdev_tx_t ieee80211_monitor_start_xmit(struct sk_buff *skb,
 		chandef = &chanctx_conf->def;
 	else if (!local->use_chanctx)
 		chandef = &local->_oper_chandef;
-	else
+	else {
+		ieee80211_tx_drop_stats(sdata, info->tid, TX_DROP_BAD_CHANCTX_CONF);
 		goto fail_rcu;
+	}
 
 	/*
 	 * Frame injection is not allowed if beaconing is not allowed
@@ -2505,8 +2588,10 @@ netdev_tx_t ieee80211_monitor_start_xmit(struct sk_buff *skb,
 	 * monitor flag interfaces used for AP support.
 	 */
 	if (!cfg80211_reg_can_beacon(local->hw.wiphy, chandef,
-				     sdata->vif.type))
+				     sdata->vif.type)) {
+		ieee80211_tx_drop_stats(sdata, info->tid, TX_DROP_MISC);
 		goto fail_rcu;
+	}
 
 	info->band = chandef->chan->band;
 
@@ -2524,8 +2609,10 @@ netdev_tx_t ieee80211_monitor_start_xmit(struct sk_buff *skb,
 	 * selected chandef above to accurately set injection rates and
 	 * retransmissions.
 	 */
-	if (!ieee80211_parse_tx_radiotap(skb, dev))
+	if (!ieee80211_parse_tx_radiotap(skb, dev)) {
+		ieee80211_tx_drop_stats(sdata, info->tid, TX_DROP_SKB_SANITY_CHECK_FAIL);
 		goto fail_rcu;
+	}
 
 	/* remove the injection radiotap header */
 	skb_pull(skb, len_rthdr);
@@ -2710,6 +2797,7 @@ static struct sk_buff *ieee80211_build_hdr(struct ieee80211_sub_if_data *sdata,
 
 	info = IEEE80211_SKB_CB(skb);
 	memset(info, 0, sizeof(*info));
+	info->tid = skb->priority & IEEE80211_QOS_CTL_TID_MASK;
 
 	/* convert Ethernet header to proper 802.11 header (based on
 	 * operation mode) */
@@ -2777,6 +2865,7 @@ static struct sk_buff *ieee80211_build_hdr(struct ieee80211_sub_if_data *sdata,
 			conf = rcu_dereference(sdata->vif.link_conf[link_id]);
 			if (unlikely(!conf)) {
 				ret = -ENOLINK;
+				ieee80211_tx_drop_stats(sdata, info->tid, TX_DROP_BAD_LINK_CONF);
 				goto free;
 			}
 
@@ -2923,11 +3012,13 @@ static struct sk_buff *ieee80211_build_hdr(struct ieee80211_sub_if_data *sdata,
 		break;
 	default:
 		ret = -EINVAL;
+		ieee80211_tx_drop_stats(sdata, info->tid, TX_DROP_UNKNOWN_VIF_TYPE);
 		goto free;
 	}
 
 	if (!chanctx_conf) {
 		if (!ieee80211_vif_is_mld(&sdata->vif)) {
+			ieee80211_tx_drop_stats(sdata, info->tid, TX_DROP_BAD_CHANCTX_CONF);
 			ret = -ENOTCONN;
 			goto free;
 		}
@@ -2970,6 +3061,7 @@ static struct sk_buff *ieee80211_build_hdr(struct ieee80211_sub_if_data *sdata,
 
 		I802_DEBUG_INC(local->tx_handlers_drop_unauth_port);
 
+		ieee80211_tx_drop_stats(sdata, info->tid, TX_DROP_UNAUTH_PORT);
 		ret = -EPERM;
 		goto free;
 	}
@@ -3029,6 +3121,7 @@ static struct sk_buff *ieee80211_build_hdr(struct ieee80211_sub_if_data *sdata,
 		head_need += local->tx_headroom;
 		head_need = max_t(int, 0, head_need);
 		if (ieee80211_skb_resize(sdata, skb, head_need, ENCRYPT_DATA)) {
+			ieee80211_tx_drop_stats(sdata, info->tid, TX_DROP_SKB_RESIZE_FAIL);
 			ieee80211_free_txskb(&local->hw, skb);
 			skb = NULL;
 			return ERR_PTR(-ENOMEM);
@@ -3087,6 +3180,7 @@ nss_mesh:
 					     sdata->name, hdr.addr1,
 					     pre_conf_link_id, link_id);
 #endif
+			ieee80211_tx_drop_stats(sdata, info->tid, TX_DROP_BAD_LINK_ID);
 			ret = -EINVAL;
 			goto free;
 		}
@@ -3655,8 +3749,10 @@ ieee80211_xmit_fast_finish(struct ieee80211_sub_if_data *sdata,
 	u8 tid = IEEE80211_NUM_TIDS;
 
 	if (!ieee80211_hw_check(&tx->local->hw, HAS_RATE_CONTROL) &&
-	    ieee80211_tx_h_rate_ctrl(tx) != TX_CONTINUE)
+	    ieee80211_tx_h_rate_ctrl(tx) != TX_CONTINUE) {
+		ieee80211_tx_drop_stats(sdata, info->tid, TX_DROP_HW_CHECK_FAIL);
 		return TX_DROP;
+	}
 
 	if (key)
 		info->control.hw_key = &key->conf;
@@ -3778,6 +3874,7 @@ void __ieee80211_xmit_fast(struct ieee80211_sub_if_data *sdata,
 	int extra_head = fast_tx->hdr_len - (ETH_HLEN - 2);
 	bool skip_tx_queue = (local->hw.perf_mode &&
 			      !ieee80211_vif_is_mesh(&sdata->vif));
+	u8 tid = skb->priority & IEEE80211_QOS_CTL_TAG1D_MASK;
 
 	skb = skb_share_check(skb, GFP_ATOMIC);
 	if (unlikely(!skb))
@@ -3794,8 +3891,10 @@ void __ieee80211_xmit_fast(struct ieee80211_sub_if_data *sdata,
 	if (unlikely(ieee80211_skb_resize(sdata, skb,
 					  max_t(int, extra_head + hw_headroom -
 						     skb_headroom(skb), 0),
-					  ENCRYPT_NO)))
+					  ENCRYPT_NO))) {
+		ieee80211_tx_drop_stats(sdata, tid, TX_DROP_SKB_RESIZE_FAIL);
 		goto free;
+	}
 
 	hdr = skb_push(skb, extra_head);
 	memcpy(skb->data, fast_tx->hdr, fast_tx->hdr_len);
@@ -3811,17 +3910,15 @@ void __ieee80211_xmit_fast(struct ieee80211_sub_if_data *sdata,
 	info->control.flags = IEEE80211_TX_CTRL_FAST_XMIT |
 			      u32_encode_bits(IEEE80211_LINK_UNSPECIFIED,
 					      IEEE80211_TX_CTRL_MLO_LINK);
+	info->tid = tid;
 
 #ifdef CPTCFG_MAC80211_DEBUGFS
 	if (local->force_tx_status)
 		info->flags |= IEEE80211_TX_CTL_REQ_TX_STATUS;
 #endif
 
-	if (hdr->frame_control & cpu_to_le16(IEEE80211_STYPE_QOS_DATA)) {
-		u8 tid = skb->priority & IEEE80211_QOS_CTL_TAG1D_MASK;
-
+	if (hdr->frame_control & cpu_to_le16(IEEE80211_STYPE_QOS_DATA))
 		*ieee80211_get_qos_ctl(hdr) = tid;
-	}
 
 	__skb_queue_head_init(&tx.skbs);
 
@@ -3873,7 +3970,7 @@ static bool ieee80211_xmit_fast(struct ieee80211_sub_if_data *sdata,
 	struct tid_ampdu_tx *tid_tx = NULL;
 	struct sk_buff *next;
 	struct ethhdr eth;
-	u8 tid = IEEE80211_NUM_TIDS;
+	u8 tid = skb->priority & IEEE80211_QOS_CTL_TAG1D_MASK;
 
 	/* control port protocol needs a lot of special handling */
 	if (cpu_to_be16(ethertype) == sdata->control_port_protocol)
@@ -3888,7 +3985,6 @@ static bool ieee80211_xmit_fast(struct ieee80211_sub_if_data *sdata,
 		return false;
 
 	if (hdr->frame_control & cpu_to_le16(IEEE80211_STYPE_QOS_DATA)) {
-		tid = skb->priority & IEEE80211_QOS_CTL_TAG1D_MASK;
 		tid_tx = rcu_dereference(sta->ampdu_mlme.tid_tx[tid]);
 		if (tid_tx) {
 			if (!test_bit(HT_AGG_STATE_OPERATIONAL, &tid_tx->state))
@@ -3902,8 +3998,10 @@ static bool ieee80211_xmit_fast(struct ieee80211_sub_if_data *sdata,
 
 	/* after this point (skb is modified) we cannot return false */
 	skb = ieee80211_tx_skb_fixup(skb, ieee80211_sdata_netdev_features(sdata), sdata->dev);
-	if (!skb)
+	if (!skb) {
+		ieee80211_tx_drop_stats(sdata, tid, TX_DROP_TX_SKB_FIXUP_FAIL);
 		return true;
+	}
 
 	skb_list_walk_safe(skb, skb, next) {
 		skb_mark_not_on_list(skb);
@@ -3999,6 +4097,7 @@ begin:
 			      !ieee80211_is_our_addr(tx.sdata, hdr->addr2,
 						     NULL)))) {
 			I802_DEBUG_INC(local->tx_handlers_drop_unauth_port);
+			ieee80211_tx_drop_stats(tx.sdata, info->tid, TX_DROP_UNAUTH_PORT);
 			ieee80211_free_txskb(&local->hw, skb);
 			goto begin;
 		}
@@ -4395,11 +4494,12 @@ void __ieee80211_subif_start_xmit(struct sk_buff *skb,
 	struct sk_buff *next;
 	int len = skb->len;
 	struct ieee80211_key *key = NULL;
-	struct ieee80211_tx_info *info;
+	struct ieee80211_tx_info *info = IEEE80211_SKB_CB(skb);
 	struct ieee80211_sub_if_data *ap_sdata;
 
 	if (unlikely(!ieee80211_sdata_running(sdata) || skb->len < ETH_HLEN)) {
-		kfree_skb(skb);
+		ieee80211_tx_drop_stats(sdata, info->tid, TX_DROP_SDATA_STATE);
+		ieee80211_free_txskb(&local->hw, skb);
 		return;
 	}
 
@@ -4412,8 +4512,10 @@ void __ieee80211_subif_start_xmit(struct sk_buff *skb,
 	    ieee80211_mesh_xmit_fast(sdata, skb, ctrl_flags))
 		goto out;
 
-	if (ieee80211_lookup_ra_sta(sdata, skb, &sta))
+	if (ieee80211_lookup_ra_sta(sdata, skb, &sta)) {
+		ieee80211_tx_drop_stats(sdata, info->tid, TX_DROP_RA_STA_FAIL);
 		goto out_free;
+	}
 
 	if (IS_ERR(sta))
 		sta = NULL;
@@ -4459,6 +4561,7 @@ void __ieee80211_subif_start_xmit(struct sk_buff *skb,
 	 */
 	skb = ieee80211_tx_skb_fixup(skb, 0, dev);
 	if (!skb) {
+		ieee80211_tx_drop_stats(sdata, info->tid, TX_DROP_TX_SKB_FIXUP_FAIL);
 		len = 0;
 		goto out;
 	}
@@ -4476,7 +4579,6 @@ void __ieee80211_subif_start_xmit(struct sk_buff *skb,
 			goto out;
 		}
 
-		info = IEEE80211_SKB_CB(skb);
 		info->sawf.nw_delay = (u32) ktime_to_us(net_timedelta(skb->tstamp));
 		if (info->flags & IEEE80211_TX_CTL_HW_80211_ENCAP) {
 		    	if (sta)
@@ -4682,6 +4784,8 @@ netdev_tx_t ieee80211_subif_start_xmit(struct sk_buff *skb,
 {
 	struct ieee80211_sub_if_data *sdata = IEEE80211_DEV_TO_SUB_IF(dev);
 	const struct ethhdr *eth = (void *)skb->data;
+	struct ieee80211_tx_info *info = IEEE80211_SKB_CB(skb);
+	bool tid_stats_disable = sdata->local->hw.tid_stats_disable;
 
 #ifdef CPTCFG_MAC80211_ATHMEMDEBUG
 	ath_update_alloc(skb, skb->truesize, __LINE__, __func__, 1);
@@ -4690,12 +4794,16 @@ netdev_tx_t ieee80211_subif_start_xmit(struct sk_buff *skb,
 #ifdef CPTCFG_MAC80211_NSS_SUPPORT
 	ieee80211_xmit_nss_fixup(skb, dev);
 #endif
-
+	if (!tid_stats_disable) {
+		skb->priority = cfg80211_classify8021d(skb, NULL);
+		ieee80211_tid_classifier(skb, sdata, false, TX_NWIFI_PKT);
+	}
 	if (likely(!is_multicast_ether_addr(eth->h_dest)))
 		goto normal;
 
 	if (unlikely(!ieee80211_sdata_running(sdata))) {
-		kfree_skb(skb);
+		ieee80211_tx_drop_stats(sdata, info->tid, TX_DROP_SDATA_STATE);
+		ieee80211_free_txskb(&sdata->local->hw, skb);
 		return NETDEV_TX_OK;
 	}
 
@@ -4814,6 +4922,7 @@ static void ieee80211_8023_xmit(struct ieee80211_sub_if_data *sdata,
 	u16 queue;
 	unsigned char *ra = ehdr->h_dest;
 	bool multicast;
+	bool tid_stats_disable = sdata->local->hw.tid_stats_disable;
 	u8 tid;
 
 	if (unlikely(!ieee80211_hw_check(&local->hw, HAS_TX_QUEUE))) {
@@ -4821,15 +4930,24 @@ static void ieee80211_8023_xmit(struct ieee80211_sub_if_data *sdata,
 		skb_set_queue_mapping(skb, queue);
 	}
 
+	if (!tid_stats_disable) {
+		skb->priority = cfg80211_classify8021d(skb, NULL);
+		tid = skb->priority & IEEE80211_QOS_CTL_TAG1D_MASK;
+	}
+
 	multicast = is_multicast_ether_addr(ra);
 
 	if (multicast && sdata->vif.type == NL80211_IFTYPE_AP_VLAN &&
-	    !atomic_read(&sdata->u.vlan.num_mcast_sta))
+	    !atomic_read(&sdata->u.vlan.num_mcast_sta)) {
+		ieee80211_tx_drop_stats(sdata, tid, TX_DROP_RA_STA_FAIL);
 		goto out_free;
+	}
 
 	if (unlikely(test_bit(SCAN_SW_SCANNING, &local->scanning)) &&
-	    test_bit(SDATA_STATE_OFFCHANNEL, &sdata->state))
+	    test_bit(SDATA_STATE_OFFCHANNEL, &sdata->state)) {
+		ieee80211_tx_drop_stats(sdata, tid, TX_DROP_SDATA_STATE);
 		goto out_free;
+	}
 
 	skb = skb_share_check(skb, GFP_ATOMIC);
 	if (unlikely(!skb))
@@ -4855,8 +4973,10 @@ static void ieee80211_8023_xmit(struct ieee80211_sub_if_data *sdata,
 	}
 
 	skb = ieee80211_tx_skb_fixup(skb, ieee80211_sdata_netdev_features(sdata), dev);
-	if (!skb)
+	if (!skb) {
+		ieee80211_tx_drop_stats(sdata, tid, TX_DROP_TX_SKB_FIXUP_FAIL);
 		return;
+	}
 
 	info = IEEE80211_SKB_CB(skb);
 	memset(info, 0, sizeof(*info));
@@ -4864,6 +4984,7 @@ static void ieee80211_8023_xmit(struct ieee80211_sub_if_data *sdata,
 	info->flags |= info_flags;
 	info->sawf.nw_delay = (u32) ktime_to_us(net_timedelta(skb->tstamp));
 	info->hw_queue = sdata->vif.hw_queue[queue];
+	info->tid = tid;
 
 	if (sdata->vif.type == NL80211_IFTYPE_AP_VLAN &&
 		!ieee80211_hw_check(&local->hw, SUPPORTS_NSS_OFFLOAD))
@@ -4943,6 +5064,7 @@ void ieee80211_8023_xmit_ap(struct ieee80211_sub_if_data *sdata,
 	info = IEEE80211_SKB_CB(skb);
 	memset(info, 0, sizeof(*info));
 	info->flags |= info_flags;
+	info->tid = skb->priority & IEEE80211_QOS_CTL_TAG1D_MASK;
 
 	if (unlikely((skb->sk &&
 		     skb_shinfo(skb)->tx_flags & SKBTX_WIFI_STATUS) ||
@@ -5003,17 +5125,23 @@ netdev_tx_t ieee80211_subif_start_xmit_8023(struct sk_buff *skb,
 	struct ieee80211_tx_info *info = IEEE80211_SKB_CB(skb);
 	struct ieee80211_tx_control control = {};
 	bool perf_mode = sdata->local->hw.perf_mode;
+	bool tid_stats_disable = sdata->local->hw.tid_stats_disable;
 
 #ifdef CPTCFG_MAC80211_ATHMEMDEBUG
         ath_update_alloc(skb, skb->truesize, __LINE__, __func__, 1);
 #endif
 
 	orig_sdata = sdata;
+
+	if (!tid_stats_disable)
+		skb->priority = cfg80211_classify8021d(skb, NULL);
 	if (likely(skb->fast_xmit && perf_mode)) {
 		if (sdata->vif.type == NL80211_IFTYPE_AP_VLAN)
 			sdata = container_of(sdata->bss,
 					     struct ieee80211_sub_if_data, u.ap);
 
+		if (!tid_stats_disable)
+			 ieee80211_tid_classifier(skb, sdata, true, TX_ETH_PKT);
 		info->control.vif = &sdata->vif;
 		info->control.flags = u32_encode_bits(IEEE80211_LINK_UNSPECIFIED,
 						      IEEE80211_TX_CTRL_MLO_LINK);
@@ -5042,7 +5170,9 @@ netdev_tx_t __ieee80211_subif_start_xmit_8023(struct sk_buff *skb,
 					      u64 *cookie)
 {
 	struct ieee80211_sub_if_data *sdata = IEEE80211_DEV_TO_SUB_IF(dev);
+	struct ieee80211_tx_info *info = IEEE80211_SKB_CB(skb);
 	struct ethhdr *ehdr = (struct ethhdr *)skb->data;
+	bool tid_stats_disable = sdata->local->hw.tid_stats_disable;
 	struct ieee80211_key *key = NULL;
 	struct sta_info *sta;
 	bool is_eapol;
@@ -5051,6 +5181,7 @@ netdev_tx_t __ieee80211_subif_start_xmit_8023(struct sk_buff *skb,
        ieee80211_xmit_nss_fixup(skb, dev);
 #endif
 	if (unlikely(!ieee80211_sdata_running(sdata) || skb->len < ETH_HLEN)) {
+		ieee80211_tx_drop_stats(sdata, info->tid, TX_DROP_SDATA_STATE);
 		kfree_skb(skb);
 		return NETDEV_TX_OK;
 	}
@@ -5058,9 +5189,11 @@ netdev_tx_t __ieee80211_subif_start_xmit_8023(struct sk_buff *skb,
 	rcu_read_lock();
 
 	if (ieee80211_lookup_ra_sta(sdata, skb, &sta)) {
+		ieee80211_tx_drop_stats(sdata, info->tid, TX_DROP_RA_STA_FAIL);
 		kfree_skb(skb);
 		goto out;
 	}
+
 	is_eapol = (sdata->control_port_protocol == ehdr->h_proto);
 
 	if (ieee80211_hw_check(&sdata->local->hw, SUPPORTS_NSS_OFFLOAD)) {
@@ -5079,6 +5212,9 @@ netdev_tx_t __ieee80211_subif_start_xmit_8023(struct sk_buff *skb,
 	if (key && (!(key->flags & KEY_FLAG_UPLOADED_TO_HARDWARE)))
 		goto skip_offload;
 
+	if (!tid_stats_disable)
+		ieee80211_tid_classifier(skb, sdata, false, TX_ETH_PKT);
+
 	if (sdata->vif.type == NL80211_IFTYPE_AP) {
 		ieee80211_8023_xmit_ap(sdata, dev, sta, key, skb, info_flags, ctrl_flags, cookie);
 		goto out;
@@ -5176,6 +5312,7 @@ static bool ieee80211_tx_pending_skb(struct ieee80211_local *local,
 			chanctx_conf =
 				rcu_dereference(sdata->vif.bss_conf.chanctx_conf);
 			if (unlikely(!chanctx_conf)) {
+				ieee80211_tx_drop_stats(sdata, info->tid, TX_DROP_BAD_CHANCTX_CONF);
 				dev_kfree_skb(skb);
 				return true;
 			}
@@ -5184,6 +5321,7 @@ static bool ieee80211_tx_pending_skb(struct ieee80211_local *local,
 		result = ieee80211_tx(sdata, NULL, skb, true);
 	} else if (info->flags & IEEE80211_TX_CTL_HW_80211_ENCAP) {
 		if (ieee80211_lookup_ra_sta(sdata, skb, &sta)) {
+			ieee80211_tx_drop_stats(sdata, info->tid, TX_DROP_RA_STA_FAIL);
 			dev_kfree_skb(skb);
 			return true;
 		}
diff --git a/net/mac80211/wep.c b/net/mac80211/wep.c
index 5c01e12..e5d9c42 100644
--- a/net/mac80211/wep.c
+++ b/net/mac80211/wep.c
@@ -299,6 +299,7 @@ ieee80211_crypto_wep_encrypt(struct ieee80211_tx_data *tx)
 	skb_queue_walk(&tx->skbs, skb) {
 		if (wep_encrypt_skb(tx, skb) < 0) {
 			I802_DEBUG_INC(tx->local->tx_handlers_drop_wep);
+			ieee80211_tx_drop_stats(tx->sdata, 0, TX_DROP_WEP);
 			return TX_DROP;
 		}
 	}
diff --git a/net/mac80211/wpa.c b/net/mac80211/wpa.c
index 21ae441..4abde7f 100644
--- a/net/mac80211/wpa.c
+++ b/net/mac80211/wpa.c
@@ -43,8 +43,10 @@ ieee80211_tx_h_michael_mic_add(struct ieee80211_tx_data *tx)
 		return TX_CONTINUE;
 
 	hdrlen = ieee80211_hdrlen(hdr->frame_control);
-	if (skb->len < hdrlen)
+	if (skb->len < hdrlen) {
+		ieee80211_tx_drop_stats(tx->sdata, info->tid, TX_DROP_SKB_SANITY_CHECK_FAIL);
 		return TX_DROP;
+	}
 
 	data = skb->data + hdrlen;
 	data_len = skb->len - hdrlen;
@@ -71,8 +73,10 @@ ieee80211_tx_h_michael_mic_add(struct ieee80211_tx_data *tx)
 		 skb_headroom(skb) < IEEE80211_TKIP_IV_LEN,
 		 "mmic: not enough head/tail (%d/%d,%d/%d)\n",
 		 skb_headroom(skb), IEEE80211_TKIP_IV_LEN,
-		 skb_tailroom(skb), tail))
+		 skb_tailroom(skb), tail)) {
+		ieee80211_tx_drop_stats(tx->sdata, info->tid, TX_DROP_SKB_SANITY_CHECK_FAIL);
 		return TX_DROP;
+	}
 
 	mic = skb_put(skb, MICHAEL_MIC_LEN);
 
-- 
2.34.1

